site/book/05-developing-functions/02-developing-in-Go.md

```diff
@@ -1,133 +1,135 @@
 
 ```go
 // main.go
 package main
 
 import (
-	"os"
+  "fmt"
+  "os"
 
-	"sigs.k8s.io/kustomize/kyaml/fn/framework"
-	"sigs.k8s.io/kustomize/kyaml/fn/framework/command"
-	"sigs.k8s.io/kustomize/kyaml/kio"
-	"sigs.k8s.io/kustomize/kyaml/yaml"
+  "github.com/GoogleContainerTools/kpt-functions-sdk/go/fn"
 )
 
+// EDIT THIS FUNCTION!
+// This is the main logic. rl is the input `ResourceList` which has the `FunctionConfig` and `Items` fields.
+// You can modify the `Items` and add result information to `rl.Results`.
+func Run(rl *fn.ResourceList) (bool, error) {
+  // Your code
+}
+
 func main() {
-	// create a struct matching the structure of ResourceList.FunctionConfig to hold its data
-	var config struct {
-		Data map[string]string `yaml:"data"`
-	}
-	fn := func(items []*yaml.RNode) ([]*yaml.RNode, error) {
-		for i := range items {
-			// set the annotation on each resource item
-			err := items[i].PipeE(yaml.SetAnnotation("myannotation", config.Data["myannotation"]))
-			if err != nil {
-				return nil, err
-			}
-		}
-		return items, nil
-	}
-	p := framework.SimpleProcessor{Filter: kio.FilterFunc(fn), Config: &config}
-	cmd := command.Build(p, command.StandaloneDisabled, false)
-  // Adds a "gen" subcommand to create a Dockerfile for building the function into a container image.
-  command.AddGenerateDockerfile(cmd)
-
-	if err := cmd.Execute(); err != nil {
-		os.Exit(1)
-	}
+  // CUSTOMIZE IF NEEDED
+  // `AsMain` accepts a `ResourceListProcessor` interface.
+  // You can explore other `ResourceListProcessor` structs in the SDK or define your own.
+  if err := fn.AsMain(fn.ResourceListProcessorFunc(Run)); err != nil {
+    os.Exit(1)
+  }
 }
 ```
 
-### Build and test the function
+The [`fn`] library provides a series of KRM level operations for [`ResourceList`]. 
+Basically, the KRM resource `ResourceList.FunctionConfig` and KRM resources `ResourceList.Items` are both converted to 
+`KubeObject` objects. You can use `KubeObject` similar as [`unstructured.Unstrucutred`].
 
-Build the go binary:
+The set-annotation function (see below) iterates the `ResourceList.Items`, finds out the `Deployment` resources and
+adds the annotation. After the iteration, it adds some user message to the `ResourceList.Results`
 
-```shell
-$ go mod tidy; go build -o my-fn .
+```go
+func Run(rl *fn.ResourceList) (bool, error) {
+    for _, kubeObject := range rl.Items {
+        if kubeObject.IsGVK("v1", "Deployment") {
+            kubeObject.SetAnnotation("config.kubernetes.io/managed-by", "kpt")
+        }
+    }
+    // This result message will be displayed in the function evaluation time. 
+    rl.Results = append(rl.Results, fn.GeneralResult("Add config.kubernetes.io/managed-by=kpt to all `Deployment` resources", fn.Info))
+    return true, nil
+}
 ```
 
-Fetch the wordpress package:
+Learn more about the `KubeObject` from the [go doc](https://pkg.go.dev/github.com/GoogleContainerTools/kpt-functions-sdk/go/fn).
```

Comment:
I suggest mention the examples and provide a link.
Normally examples in go doc live along with each type definition. But it's not the case since how we organise them. But it's not very easy to discover it since it's at the very bottom.

Reply:
yeah, that's true. I like your original in-line go doc for the kyaml RNode. Anyway, I'll update this to point to the examples.

Reply:
I added the go-doc examples in the "Next Step" section, and an extended reading to show other KRM authoring use cases, and keep this link pointing to the sdk which lists all the functions KubeObject provides.

Reply:
> I like your original in-line go doc for the kyaml RNode.

Yes, me too. But we can't leak the k8s dependencies to the main `go/fn` module.
It's a nightmare if `go/fn` depends on one k8s version while you also need something else that depends on a different k8s version.